[Video] Add Reverse Cycle Subtitles action

bjsvedin opened the pull request:
Added ACTION_REVERSE_CYCLE_SUBTITLE action. Made cycling subtitles while disable possible. Modifed when the Info toast for SHOW_SUBTITLE and NEXT/CYCLE_SUBTITLE will show and what it displays.

## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
I added the action Reverse Cycle Subtitle which behaves like Cycle Subtitle, but in the opposite direction. I also made it possible to cycle subtitles while subtitles were disabled.

<!--- Describe your change in detail here. -->
When subtitles are enabled, Cycle behaves the same as before, and Reverse Cycle behaves the same as Cycle, only in the opposite direction. When making cycling through subtitles when subtitles are disabled possible, I needed to modify the Info window's text to make it clear what action is being taken. If subtitles are disabled, and you cycle, the title will be "Subtitles(Disabled)", with the language/name as the message. I also did this modification to the Show Subtitles action on disabling. I did not modify the Info window text when Next Subtitle disables subtitles.

Next Subtitle behavior did not change. I wanted to add a Previous Subtitle action that behaved like Next Subtitle, but in the opposite direction. To get those two actions to play nicely with each other at the edges of the lists turned out to be either not possible, or far too much work to be worth it, so I scrapped that idea. Reverse Cycle meets the requirements I was looking for.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
I wanted the ability to cycle BOTH directions through the subtitles. There are times that a video has over 40 different subtitles, and it is tedious to select the proper subtitle if I accidentally go one step too far when turning on subtitles. I either needs to spam the next button, or pull up the subtitle menu to manually set it. 
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
My tests were manual and brute forced. I setup my joystick.xml to have all subtitle actions available to me. I would build and install Kodi with my changes and load up a video with at least 2 subtitles. I then tested cycling, enabling, disabling, and using next, to confirm that the behavior is as expected. The subtitle cycle wrap worked as expected in both directions, and was not altered for Next Subtitle.

<!--- Include details of your testing environment, and the tests you ran to -->
I develop, build, and test on vanilla Windows 11 machines. I do have two machines I test on.

<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
This won't affect the behaviors users are already use to. Next Subtitle will behave the same, and so with Cycle Subtitle when subtitles are enabled. The message they see will be slightly altered when subtitles are disabled, which should make it more explicit what the state of the subtitles and selected subtitles are. This will then add functionality they did not have before, which is cycling subtitles when they are disabled.
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki (Document the new action available)
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

bjsvedin committed:
Added ACTION_REVERSE_CYCLE_SUBTITLE action. Made cycling subtitles while disable possible. Modifed when the Info toast for SHOW_SUBTITLE and NEXT/CYCLE_SUBTITLE will show and what it displays.

bjsvedin commented:
This is a draft PR because I don't know what to do with this line

`ADDON_ACTION_REVERSE_CYCLE_SUBTITLE = 9999999999, // What do I do with this value? Do I use the next available or shift everything?`

There are hundreds of values in there. I do not know what value to assign it. for my testing this existing number works.

neo1973 commented on xbmc/video/PlayerController.cpp:
```diff
@@ -74,24 +74,23 @@ bool CPlayerController::OnAction(const CAction &action)
+
+        if (!subsOn)
+          caption =
+              StringUtils::Format("{}({})", caption, g_localizeStrings.Get(1223)); /*Disabled*/
```
I think there's a space missing:
```suggestion
              StringUtils::Format("{} ({})", caption, g_localizeStrings.Get(1223)); /*Disabled*/
```

neo1973 reviewed (changes requested):
In general a useful addition but I want to suggest some changes:

1. Please split the change to the notification and the `ACTION_REVERSE_CYCLE_SUBTITLE` change into separate commits
2. I think `ACTION_REVERSE_CYCLE_SUBTITLE` and `ACTION_CYCLE_SUBTITLE` are similar enough that the cycling should be moved into a common member function where a parameter decides if the cycling is forwards or backwards. `ACTION_NEXT_SUBTITLE` would also use this function and only add the enable/disable behaviour around it.

bjsvedin commented:
>     1. Please split the change to the notification and the `ACTION_REVERSE_CYCLE_SUBTITLE` change into separate commits

Sure, I can do that, though I don't see a clear line for where that split needs to be. There are actually three things done here: 

1. Added ACTION_REVERSE_CYCLE_SUBTITLE, 
2. Made cycling subtitle while subtitles disabled possible
3. Modified the notification text.

The notification change was made because of the cycle while disabled change was made, and it was suddenly not obvious what the state of the subtitles showing was. I think the best place to split these changes are between 1 and 2, but maybe it makes sense to have three separate commits.

>     2. I think `ACTION_REVERSE_CYCLE_SUBTITLE` and `ACTION_CYCLE_SUBTITLE` are similar enough that the cycling should be moved into a common member function where a parameter decides if the cycling is forwards or backwards. `ACTION_NEXT_SUBTITLE` would also use this function and only add the enable/disable behaviour around it.

I looked into doing this. Would you be ok if i simply merged the switch case 'ACTION_REVERSE_CYCLE_SUBTITLE' into the other on that holds 'ACTION_CYCLE_SUBTITLE' and 'ACTION_NEXT_SUBTITLE', or should these actions be separate, with the common use function? I ask because the only difference is the direction. Notifications, and visibility behaviors are the same.

Ch1llb0 commented:
This is a great addition! Was missing this when adding a subtitle cycle option into my skin...

bjsvedin commented:
I currently don't have time to work on this PR. I like the new feature, but it's not necessary for me. I am going to close this PR for the time being. I may one day get back to this if someone else hasn't already done it.
